Feature/pf 896 content negotiation

alexshook opened the pull request:
https://salido.atlassian.net/browse/PF-896

alexshook committed:
Require the new services directory so we can use services

alexshook committed:
WIP begin expected Content Negotiation JSON API compliance

alexshook committed:
Also add the services file to require services

alexshook committed:
Add service to validate JSON API content negotiation

- According to JSON API v1.0, this checks the headers for compliance with
the specifications.
- It currently returns 406 and 415 when appropriate.
- If all checks pass, it returns true.
- If the api Accept-Version header is wrong, it returns false.
- Returning true, false, or error seems like one too many things to
return from a service, and I'd rather have it return true or error.
- In the case of the Accept-Version header being wrong (returns false),
I think it could also return 406. Or 500, but I think 406 is slightly
better.

alexshook committed:
Rubocop fixes

alexshook commented on lib/grape/jsonapi/services/content_negotiator.rb:
```diff
@@ -0,0 +1,50 @@
+          new(*args).run
+        end
+
+        def run
```
@chandeeland I'm not super pleased with this returning `true`, `false`, or error. I think `true` or error would be better, I left a long commit message about it here: https://github.com/chandeeland/grape-jsonapi_entity/pull/15/commits/d90f91871332ff9d952d421e133cc15835613073.

We could also try moving the error handling entirely into the API, but I think it's better here. 

Also this might be full of holes so please poke it a few times. 

For reference: 
- JSON API specifications on content negotiation: http://jsonapi.org/format/#content-negotiation
- Grape documentation on using the Accept-Version header https://github.com/ruby-grape/grape#accept-version-header

chandeeland commented on lib/grape/jsonapi/services/content_negotiator.rb:
```diff
@@ -0,0 +1,50 @@
+      class ContentNegotiator
+        VALID_MEDIA_TYPE = 'application/vnd+json'.freeze
+
+        ERRORS = {
```
lets promote these to full Error objects

chandeeland commented on lib/grape/jsonapi/services/content_negotiator.rb:
```diff
@@ -0,0 +1,50 @@
+
+        def run
+          return ERRORS[:unsupported_media] unless valid?(content_type)
+          return ERRORS[:not_acceptable] unless valid?(accept_header)
```
lets just raise instead of returning

chandeeland commented on lib/grape/jsonapi/services/content_negotiator.rb:
```diff
@@ -0,0 +1,50 @@
+          return ERRORS[:unsupported_media] unless valid?(content_type)
+          return ERRORS[:not_acceptable] unless valid?(accept_header)
+
+          valid_accept_version_header?
```
lets remove all concept of versioning and accept-version header from this service object and only focus on the content-type header

chandeeland reviewed (dismissed):
no version 💩

alexshook committed:
Add test for valid media type constant

Oops I forgot it :(

alexshook committed:
Remove the checks for Accept-Version header

alexshook committed:
Add ContentNegotiationError to exceptions module

alexshook committed:
Use the module convention for exceptions now that there is more than one

alexshook committed:
Split Content Negotiation exception into individual exceptions

This will make it easier for us to return the different http status
codes in the API when we rescue the exceptions.

alexshook committed:
Raise exceptions instead of returning error hash

alexshook committed:
Rubocop fixes

inactdev commented on lib/grape/jsonapi/services/content_negotiator.rb:
```diff
@@ -0,0 +1,41 @@
+        attr_reader :content_type, :accept_header
+
+        def valid?(header)
+          VALID_MEDIA_TYPE == header || header.nil?
```
Valid accept header is something like `application/vnd.geoffrey-v20170505+json`, no ?

alexshook replied:
It's not JSON API compliant, so no. Instead of versioning in the Accept header we're now using the Accept-Version header.

inactdev replied:
According to [this](https://github.com/alexshook/grape-jsonapi_entity/blob/11cfab8fd05de10a699c685f5c3adab69ac1743e/lib/grape/jsonapi/services/content_negotiator.rb#L9) it looks like we're pulling the `Accept` header, not `Accept-Version`.  Also, tests or it didn't happen lol ;)

inactdev reviewed (changes requested):
You're only testing the valid `Content-Type` case here.  Shouldn't you test the valid `Accept` case also ?

alexshook commented:
@aperezmontan the Accept header is being tested-- it raises 406 if invalid.

alexshook commented:
Oh wait also I forgot to bump the version... is this a major or minor?

alexshook commented:
@aperezmontan yes, we do ask for the Accept header because we check it according to JSON API content negotiation specifications http://jsonapi.org/format/#content-negotiation. We do not check version in the gem because Grape handles that in the API. If you want to check the version in the gem, you'll need to fight it out with @chandeeland, who previously told me to remove it. 

You're going to have to be more specific about what you want in tests because the Accept header is being tested so I don't know what you're asking for. 

It also might not matter because I need to make a change anyway. I want to modify the arguments we pass to this gem to cut down on the number of points that could require change.

alexshook committed:
Better descriptions for 406 and 415 errors

alexshook committed:
Pass Accept and Content-Type headers directly to service

- This allows us to just check what they are and that they're valid
instead of parsing the headers.
- We'll do that work in the API.

alexshook committed:
Fix typo :scream:

alexshook commented:
@aperezmontan can you re-review? I made changes so dismissed the old ones. I changed the way that we pass data into the gem and squashed down the test file a bit, which I think is clearer now (I mean I think so, you might not 🤣 ).

alexshook committed:
Make content negotiation service args required

inactdev commented on lib/grape/jsonapi/services/content_negotiator.rb:
```diff
@@ -0,0 +1,41 @@
+        attr_reader :accept_header, :content_type
+
+        def valid?(header)
+          VALID_MEDIA_TYPE == header || header.nil?
```
Sorry, should've asked this before... So no header is valid ?

inactdev replied:
@chandeeland says make it fail on `header.nil?`

alexshook replied:
😆  http://weknowmemes.com/generator/uploads/generated/g1366824328707416474.jpg

alexshook committed:
Do not allow nil Content-Type

- JSON API doesn't require this, but we're going to require that a
Content-Type be sent with all requests, regardless of type.
- So now we're JSON API compliant+.
- The Accept header is still optional.

chandeeland reviewed (changes requested):
bump version

add a line to changelog

maybe add a section to readme explaining use of this service object

alexshook committed:
Bump minor version

alexshook committed:
Add changelog description for 0.7.0

alexshook committed:
Add description of Content Negotiation to README
